<a href="https://colab.research.google.com/github/russelfordguinoo-oss/Russel-Repository-/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/russelfordguinoo-oss/Russel-Repository-/blob/main/work/notebooks/w01_research_question.ipynb)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

I chose the Refresh / Content Opportunity Scoring lane. I want to identify strong pages by looking at their clicks, sessions, and performance trend stability, then use those pages as benchmarks for comparing other pages. The goal is to help a content team prioritize which pages should be reviewed when they have limited optimization time.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

Decision

The content team needs to decide which pages should be prioritized for review based on how their observed performance compares with strong benchmark pages.

 Action

The team can compare a candidate page with strong pages and decide what content improvements, if any, are appropriate.

Cost of a wrong call

Your main concern is missing a page that actually needs improvement. A flagged page still requires human review before deciding whether improvement is appropriate.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [28]:
!git clone https://github.com/russelfordguinoo-oss/Russel-Repository.git

Cloning into 'Russel-Repository'...
fatal: could not read Username for 'https://github.com': No such device or address


In [29]:
import os

for root, dirs, files in os.walk("/content/Russel-Repository/data/raw"):
    for file in files:
        print(os.path.join(root, file))

In [30]:
import pandas as pd

url = "https://raw.githubusercontent.com/russelfordguinoo-oss/Russel-Repository-/refs/heads/main/data/raw/content_refresh_anonymized.csv"

df = pd.read_csv(url)

print("Rows:", len(df))
print("Columns:", df.columns.tolist())

Rows: 30000
Columns: ['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


In [31]:
print(df.shape)
print(df.columns.tolist())

(30000, 44)
['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


In [32]:
print(df.head(3).to_string())

             content_id          client_id  search_volume  competition competition_level   cpc     content_type    main_intent  word_count  char_count provider_used              model_used  impressions_90d  clicks_90d  pageviews_90d  sessions_90d  users_90d  engaged_sessions_90d  ai_sessions_90d  scroll_events_90d  days_with_impressions  days_with_sessions  impressions_last_30d  clicks_last_30d  sessions_last_30d  impressions_prev_30d  clicks_prev_30d  sessions_prev_30d  content_age_days age_tier  age_tier_order  days_since_last_update freshness_tier word_count_tier char_count_tier   ctr  avg_position  engagement_rate  scroll_rate  ai_traffic_pct impression_tier position_tier trend_direction  trend_pct
0  content_304f48230142  client_f369cb89fc           10.0         0.67              HIGH  2.05  keyword article  transactional      3221.0     20457.0           NaN        gemini-2.5-flash             3803          29             22            17         16                     1         

In [33]:
# Show columns related to our research question
for col in df.columns:
    if any(word in col.lower() for word in [
        "click", "session", "engag", "trend", "30d"
    ]):
        print(col)

clicks_90d
sessions_90d
engaged_sessions_90d
ai_sessions_90d
days_with_sessions
impressions_last_30d
clicks_last_30d
sessions_last_30d
impressions_prev_30d
clicks_prev_30d
sessions_prev_30d
engagement_rate
trend_direction
trend_pct


In [34]:
# Show columns related to our research question
for col in df.columns:
    if any(word in col.lower() for word in [
        "click", "session", "engag", "trend", "30d"
    ]):
        print(col)

clicks_90d
sessions_90d
engaged_sessions_90d
ai_sessions_90d
days_with_sessions
impressions_last_30d
clicks_last_30d
sessions_last_30d
impressions_prev_30d
clicks_prev_30d
sessions_prev_30d
engagement_rate
trend_direction
trend_pct


In [35]:
print("90th percentile clicks:",
      df["clicks_90d"].quantile(0.90))

print("90th percentile sessions:",
      df["sessions_90d"].quantile(0.90))

90th percentile clicks: 32.0
90th percentile sessions: 88.0


In [36]:
print(df["trend_pct"].describe())

count    26612.000000
mean        -4.785969
std        473.861780
min       -100.000000
25%        -62.600000
50%        -33.500000
75%          0.000000
max      44900.000000
Name: trend_pct, dtype: float64


In [37]:
print(df["trend_direction"].value_counts())

trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64


In [38]:
strong_pages = df[
    (
        (df["clicks_90d"] >= df["clicks_90d"].quantile(0.90))
        |
        (df["sessions_90d"] >= df["sessions_90d"].quantile(0.90))
    )
    &
    (df["trend_direction"] == "stable")
]

print("Strong benchmark pages:", len(strong_pages))
print("Percentage of all pages:", len(strong_pages) / len(df) * 100)

Strong benchmark pages: 1526
Percentage of all pages: 5.086666666666666


In [39]:
print(
    strong_pages[
        [
            "content_id",
            "clicks_90d",
            "sessions_90d",
            "trend_direction",
            "trend_pct"
        ]
    ].head(10)
)

               content_id  clicks_90d  sessions_90d trend_direction  trend_pct
3    content_331d6c4de07b          58            78          stable      -13.8
10   content_d8ee6cc6d642         324           326          stable       19.0
55   content_a519180b7a3f          60           290          stable       -8.3
74   content_1e53ea7dbee8           5           133          stable      -12.4
104  content_09afda832f06         236           533          stable       -2.3
121  content_73bdd1706363          38            48          stable       -3.1
124  content_02950673a353         177           174          stable        8.6
133  content_09a81e8d67e2         399           449          stable        6.9
194  content_54514ebf52f9         543           426          stable        3.9
214  content_6f2194f42b53           4           239          stable        1.4


In [40]:
other_pages = df.loc[~df["content_id"].isin(strong_pages["content_id"])]

comparison = pd.DataFrame({
    "Strong benchmark pages": [
        strong_pages["clicks_90d"].mean(),
        strong_pages["sessions_90d"].mean(),
        strong_pages["trend_pct"].mean()
    ],
    "Other pages": [
        other_pages["clicks_90d"].mean(),
        other_pages["sessions_90d"].mean(),
        other_pages["trend_pct"].mean()
    ]
}, index=["Average clicks (90d)", "Average sessions (90d)", "Average trend %"])

print(comparison)

                        Strong benchmark pages  Other pages
Average clicks (90d)                120.159240    10.520369
Average sessions (90d)              210.324377    27.781274
Average trend %                      -3.867693    -4.841828


The starter dataset contains 3,000 pages. Using my initial benchmark rule — top 10% in clicks or sessions and a stable trend — 152 pages (5.1%) qualify as strong benchmark pages. This gives me a manageable group of pages to study for common patterns in clicks, sessions, and performance trend stability.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

These patterns are associated with strong-performing pages and may be useful as benchmarks. I cannot claim that these patterns cause better performance or that changing a weaker page will definitely improve it.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.